Add automatted tests for Nav block editable list view

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Adds a series of tests to provide high level coverage for the editable list view in the Navigation block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently this feature has no test coverage. These tests provide that.

Moreover, the `<OffcanvasEditor>` which currently powers the editable list view in the Nav block is [due to be refactor to use the standard `<ListView>` component](https://github.com/WordPress/gutenberg/pull/49417). This refactoring would be a lot easier if the current functionality of the editable list view were covered by tests.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds Playwright e2e tests to cover the top level scenarios:

- does it show a list view
- does the list view reflect the structure of menu items
- can you add/edit/remove menu items
- can you add a submenu

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Run

```
npm run test:e2e:playwright -- -g "List view editing" 
```

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

getdave committed:
Scaffold initial test

getdave committed:
Assert on menu structure

getdave committed:
Refine structure test filters

getdave committed:
Assert using appender to add menu items

getdave committed:
Improve assertions to check the type of the link being inserted

getdave committed:
Improve asserting on Page and Custom links being 1st and 2nd results

getdave committed:
Test block removal

getdave committed:
Make test describe blocks consistent with other tests

getdave committed:
Add test for editing items

getdave committed:
Test for adding submenus

MaggieCabrera commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+
+		// The options menu button is a sibling of the menu item gridcell.
+		const firstItemOptions = firstMenuItem
+			.locator( '..' ) // parent selector.
```
TIL! neat

getdave replied:
Same. Turns out reading the docs is useful 😆

kevin940726 replied:
Nit: This technique assumes the structure of the button to be a direct child of the row, which might change as an implementation detail? I would probably refactor this to select the `row` instead in the first place:


```js
const firstMenuRow = page.getByRole( 'row' )
  .filter({ has: firstMenuItem });

const firstItemOptions = firstMenuRow
  .getByRole( 'button', {
    name: 'Options for Page Link block',
  } );
```

Isn't critical though!

getdave replied:
For whatever reason that didn't work. It ended up producing

```
getByRole( 'treegrid', { name: 'Block navigation structure' } )
	.getByRole( 'row' )
	.filter( {
		has: getByRole( 'treegrid', { name: 'Block navigation structure' } )
			.getByRole( 'gridcell', { name: 'Page Link link' } )
			.filter( { hasText: 'Block 1 of 2, Level 1' } ),
	} )
	.getByRole( 'button', { name: 'Options for Page Link block' } );
```

MaggieCabrera commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+
+		// Grab the text from the first result so we can check it was inserted.
+		const firstResultText = await firstResult
+			.locator( '.block-editor-link-control__search-item-title' ) // this is the only way to get the label text without the URL.
```
I tried to find a different way to target this to [avoid the CSS selector](https://playwright.dev/docs/best-practices#prefer-user-facing-attributes-to-xpath-or-css-selectors) but I couldn't :S

getdave replied:
Nope. There isn't a way which probably points to a flaw in the markup of the component. It's due an a11y review as part of https://github.com/WordPress/gutenberg/issues/49091

MaggieCabrera replied:
maybe add a comment referring that PR so we can update the test once that gets merged?

talldan replied:
I'm not sure there's an issue with the markup. It's a tricky one to solve.

You could possibly do something like this pseudocode:
```js
const expectedPageTitle = 'My sample page';
await linkControl.search( expectedPageTitle );
await linkControl.selectOption( expectedPageTitle );
await expect( listView ).toHaveBlock( expectedPageTitle );
```

Have the test select a known page title, rather than trying to read the text from the option.

It might also be good to assert that the URL added is correct.

kevin940726 replied:
I would agree with @MaggieCabrera to add a comment referencing that PR, but it's non-blocking 👍 .

getdave replied:
> Have the test select a known page title, rather than trying to read the text from the option.

That might be better.

getdave replied:
Only thing is if we ever change the implementation of "Recently Added" results in Link UI the test would start failing.

That's why I wanted to get the first result's text. I can extract to a helper on the utility class if that helps.

getdave replied:
I added a `getSearchResultText` method to the `LinkControl` test util class.

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+
+		await editor.openDocumentSettingsSidebar();
+
+		const listViewTab = page.getByRole( 'tab', {
+			name: 'List View',
+		} );
+
+		await listViewTab.click();
```
This could possibly be removed, as I think it's already the default active tab, and it's good for the test to fail if it's not.

getdave replied:
Good point. I'll change this to assert that the tab is visible and selected. That way if it fails the reason will be obvious.

getdave replied:
✅ I removed all the clicks. Left in a single test which asserts on visibility of the tab and thus that one will fail if the tab isn't preselected and provide cover for all the other tests.

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
 } );
+
+test.describe( 'List view editing', () => {
+	test( 'it should show a list view in the inspector controls', async ( {
```
```suggestion
	test( 'shows list view in the inspector controls', async ( {
```

In most frameworks `test` is just an alias for `it`, so the word `it` doesn't need to be in the description itself. I always think removing the word 'should' makes the description shorter and easier to read, so also suggesting that.

Maybe I should add some of this to the docs.

getdave replied:
✅

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+		} );
+
+		// Check the structure of the individual menu items matches the one that was created.
+		await expect(
+			listView
+				.getByRole( 'gridcell', {
+					name: 'Page Link link',
+				} )
+				.filter( {
+					hasText: 'Block 1 of 2, Level 1', // proxy for filtering by description.
+				} )
+				.getByText( 'Top Level Item 1' )
+		).toBeVisible();
+
+		await expect(
+			listView
+				.getByRole( 'gridcell', {
+					name: 'Submenu link',
+				} )
+				.filter( {
+					hasText: 'Block 2 of 2, Level 1', // proxy for filtering by description.
+				} )
+				.getByText( 'Top Level Item 2' )
+		).toBeVisible();
+
+		await expect(
+			listView
+				.getByRole( 'gridcell', {
+					name: 'Page Link link',
+				} )
+				.filter( {
+					hasText: 'Block 1 of 1, Level 2', // proxy for filtering by description.
+				} )
+				.getByText( 'Test Submenu Item' )
+		).toBeVisible();
```
This is really nice. I wonder if there's a way to ensure that the blocks are in the right order, maybe by selecting an `nth` row in the table as part of the selector.

It might also be good to port some of these tests over to the main List View once this PR is in, perhaps with groups and other blocks.

getdave replied:
> I wonder if there's a way to ensure that the blocks are in the right order, maybe by selecting an nth row in the table as part of the selector.

The tricky part is getting a list of gridcells for the blocks but _not_ including other `gridcells`. If you have that then you can do `nth()` and assert on that.

I spent far too long trying to do that with @kevin940726 yesterday so ended up settling on this route.

draganescu replied:
This is pretty cool!

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+			} )
+		).toBeFocused();
+
+		const blockResults = page.getByRole( 'listbox', {
+			name: 'Blocks',
+		} );
+
+		await expect( blockResults ).toBeVisible();
+
+		const blockResultOptions = blockResults.getByRole( 'option' );
+
+		// Expect to see the Page Link and Custom Link blocks as the nth(0) and nth(1) results.
+		// This is important for usability as the Page Link block is the most likely to be used.
+		await expect( blockResultOptions.nth( 0 ) ).toHaveText( 'Page Link' );
+		await expect( blockResultOptions.nth( 1 ) ).toHaveText( 'Custom Link' );
+
+		// Select the Page Link option.
+		const pageLinkResult = blockResultOptions.nth( 0 );
+		await pageLinkResult.click();
```
For some of these operations that are a bit more complex, it might help to include a utility class in the test file. Possibly one for LinkControl and one for ListView itself.

There's an example here:
https://github.com/WordPress/gutenberg/blob/9ee0f4b81b205cc43f1645ae059240022e8de0d1/test/e2e/specs/site-editor/style-variations.spec.js#L215-L234

And here's where it gets configured:
https://github.com/WordPress/gutenberg/blob/9ee0f4b81b205cc43f1645ae059240022e8de0d1/test/e2e/specs/site-editor/style-variations.spec.js#L6-L10

I think it could also be worth considering moving the block's List View tests into a separate file to keep things tidy.

getdave replied:
I haven't made any utilities yet. Curious to try that.

getdave replied:
I made one for Link Control. 

> and one for ListView itself.

I decided this was too much for this PR. Going to do a follow up.

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+			content: `<!-- wp:navigation-link {"label":"Top Level Item 1","type":"page","id":250,"url":"http://localhost:8888/quod-error-esse-nemo-corporis-rerum-repellendus/","kind":"post-type"} /-->
+			<!-- wp:navigation-submenu {"label":"Top Level Item 2","type":"page","id":250,"url":"http://localhost:8888/quod-error-esse-nemo-corporis-rerum-repellendus/","kind":"post-type"} -->
+				<!-- wp:navigation-link {"label":"Test Submenu Item","type":"page","id":270,"url":"http://localhost:8888/et-aspernatur-recusandae-non-sint/","kind":"post-type"} /-->
+			<!-- /wp:navigation-submenu -->`,
```
It'd be good to store the content in a variable, as it's repeated in lots of tests.

getdave replied:
😆 Yes. It is rather verbose. I could variablise it as well.

getdave replied:
✅

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+			} );
+
+		// Focus the node to reveal the "3 dots" options menu button.
+		const firstMenuItemAnchor = listView.getByRole( 'link', {
+			name: 'Top Level Item 1',
+			includeHidden: true,
+		} );
+		await firstMenuItemAnchor.focus();
```
I'm not sure it needs to be focused first. The dropdown buttons in ListView are always present and can be clicked on, but they're set to `opacity: 0` until the row is hovered or focused.

getdave replied:
I think I had problems getting the right selector and maybe added the focus to be sure it wasn't the visibility causing me problems. I'll try it without.

getdave replied:
I tried and the `.focus()` wasn't needed. I removed all instances.

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+		await firstMenuItemAnchor.focus();
+
+		// The options menu button is a sibling of the menu item gridcell.
+		const firstItemOptions = firstMenuItem
+			.locator( '..' ) // parent selector.
+			.getByRole( 'button', {
+				name: 'Options for Page Link block',
+			} );
```
An alternative could be to have the test delete the submenu block, as it's the only one that has a button named 'Options for Submenu block'. It would save having to select the gridcell, then the parent, then finally the dropdown menu.

getdave replied:
Yeh. I guess that would be less verbose. I don't really like the parent selector bit anyway as it's not perceivable to the user.

getdave replied:
Refactored ✅

talldan commented on test/e2e/specs/editor/blocks/navigation.spec.js:
```diff
@@ -317,3 +317,480 @@ test.describe( 'Navigation block', () => {
+			} )
+		).toBeVisible();
+
+		await expect(
+			blockSettings
+				.getByRole( 'tabpanel', {
+					name: 'Settings',
+				} )
+				.getByRole( 'heading', {
+					name: 'Link Settings',
+				} )
+		).toBeVisible();
```
It might be good to edit the label in this test, and then check that the change is reflected in List View.

getdave replied:
Yes actually that's a good point. That is something that needs to be asserted on.

getdave replied:
Updated the test to amend the label and assert on the change taking effect.

talldan reviewed (commented):
Looks good. Nice work on making them so comprehensive.

I think it would be good to add a ListView utils class, as some of the locators for gridcells span multiple lines and are repeated a lot. Test readability is one of the main things that keeps them maintainable, IMO.

It could be done after the initial merge though, as it would be good to get this in before #49417 (as was mentioned on that PR).

kevin940726 reviewed (approved):
Nice! LGTM! 👍  Thanks! ❤️

getdave committed:
Remove uneccessary test wording

getdave committed:
Remove unecessary click on List View tab

Should be visible by default. Left one test which will fail if the tab isn’t automatically pre-selected.

getdave committed:
Extract blocks to a var

getdave committed:
Add Link Control test helper and extract relevant portions of tests

getdave committed:
Extract util for getting result text

getdave committed:
Remove unneeded focus of item to trigger options menu

getdave committed:
Avoid overly complex selectors by targetted Submenu for item removal test

getdave committed:
Remove comment

getdave committed:
Test editing a menu item actually updates in the list view

getdave committed:
Remove .only
